### PR TITLE
Support Aliyun DNS

### PR DESCRIPTION
**Pull request by loganhz**
[#9382](https://github.com/rancher/rancher/issues/9382)

### CODE REVIEW

File: providers/alidns/alidns.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to describe root domain name for '%s': %v", a.rootDomainName, err)
+	}
+
+	logrus.Infof("Configured %s with zone '%s'", a.GetName(), a.rootDomainName)
+	return nil
+}
+
+func (*AlidnsProvider) GetName() string {
```
Comment (yasker) [inline providers/alidns/alidns.go]: Keep it consistent. `func (a *AlidnsProvider) ...`

File: providers/alidns/alidns.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+func (a *AlidnsProvider) HealthCheck() error {
+	_, err := a.client.DescribeDomainInfo(&api.DescribeDomainInfoArgs{
+		DomainName: a.rootDomainName,
+	})
+	return err
+}
+
+func (a *AlidnsProvider) AddRecord(record utils.DnsRecord) error {
+	for _, rec := range record.Records {
+    r := a.prepareRecord(record, rec)
```
Comment (yasker) [inline providers/alidns/alidns.go]: Run `go fmt`.

File: providers/alidns/alidns.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    _, err := a.client.AddDomainRecord(r)
```
Comment (yasker) [inline providers/alidns/alidns.go]: Use the pattern.

File: providers/alidns/alidns.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+			records = append(records, record)
+		}
+	}
+
+	return records, nil
+}
+
+func (a *AlidnsProvider) parseName(record utils.DnsRecord) string {
+	name := strings.TrimSuffix(record.Fqdn, fmt.Sprintf(".%s.", a.rootDomainName))
+	return name
```
Comment (yasker) [inline providers/alidns/alidns.go]: return the result directly.

**Comment by loganhz** [comment]
@yasker fixed 
thanks a lot for your comment

**Comment by yasker** [comment]
@loganhz 

The idea way is you should fold the new commits into the old one regarding the logic code change. You can use `git rebase -i` for it, then force push it to update the PR.

Remember to back up your code first. Force push is always risky.

**Comment by loganhz** [comment]
@yasker done 
thanks again

File: providers/alidns/alidns.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func (a *AlidnsProvider) RemoveRecord(record utils.DnsRecord) error {
+	records, err := a.findRecords(record)
+	if err != nil {
+		return err
+	}
+
+	for _, rec := range records {
+		_, err := a.client.DeleteDomainRecord(&api.DeleteDomainRecordArgs{
```
Comment (yasker) [inline providers/alidns/alidns.go]: Should be in the same line as `if`.

File: providers/alidns/alidns.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+func (a *AlidnsProvider) parseName(record utils.DnsRecord) string {
+	return strings.TrimSuffix(record.Fqdn, fmt.Sprintf(".%s.", a.rootDomainName))
+}
+
+func (a *AlidnsProvider) prepareRecord(record utils.DnsRecord, rec string) *api.AddDomainRecordArgs {
+	name := a.parseName(record)
+	return &api.AddDomainRecordArgs{
+		DomainName: a.rootDomainName,
+		RR:         name,
```
Comment (yasker) [inline providers/alidns/alidns.go]: Just move the `a.parseName` here

**Comment by yasker** [comment]
Hi @janeczku 

Can you review it as well?

File: providers/alidns/alidns.go (hunk 5 above)
Comment (loganhz) [inline providers/alidns/alidns.go]: done

File: providers/alidns/alidns.go (hunk 6 above)
Comment (loganhz) [inline providers/alidns/alidns.go]: moved! thanks

**Comment by loganhz** [comment]
@janeczku  done
thanks a lot!

**Comment by yasker** [comment]
@janeczku Does it look good now?

**Comment by janeczku** [comment]
Thanks for changes @loganhz! In case you were intending to create a Aliyun DNS catalog template this would be a good starting point: https://github.com/rancher/community-catalog/tree/master/infra-templates/digitalocean-dns/3.

**Pull request merged**